### PR TITLE
drivers: apic_timer: move to periodic mode

### PR DESCRIPTION
**Pull request by npitre**
This driver is impossible to make time-accurate using single-shot
mode. Time accuracy may be obtained only by using periodic mode, meaning
it is not tickless capable either. Let's simplify the code by only
supporting periodic mode and strip out the TSC stuff. Any hardware with
TSC capability should use the apic_tsc driver instead.

There is apparently some hardware out there with a TSC but without the
ability to set a deadline comparator. In such case it would be much
cleaner to add the ability to use the APIC ICR in one-shot mode as a 
fallback IRQ source to the apic_tsc driver instead of relying on the 
apic_timer driver.

### CODE REVIEW

File: drivers/timer/apic_timer.c [hunk 1]
```diff
@@ -31,193 +27,63 @@ BUILD_ASSERT(!IS_ENABLED(CONFIG_SMP), "APIC timer doesn't support SMP");
...
-
-	cycles = cached_icr;
-	cached_icr = MAX_TICKS * CYCLES_PER_TICK;
-	total_cycles += cycles;
-	ticks = (total_cycles - last_announcement) / CYCLES_PER_TICK;
-	last_announcement = total_cycles;
-	k_spin_unlock(&lock, key);
-	sys_clock_announce(ticks);
-}
+BUILD_ASSERT(CYCLES_PER_TICK >= 1, "APIC timer: bad CYCLES_PER_TICK");
```
Comment (andyross) [inline drivers/timer/apic_timer.c:46]: Might want a BUILD_ASSERT(!defined(CONFIG_TICKLESS_KERNEL)) given that support is being removed, for the benefit of any out-of-tree apps that might be using it.

**Review by andyross — COMMENTED** [review]
Some notes, haven't had time for a full review.  No doubt lots of bikeshedding will happen over the feature removal, but just for the record this looks fine to me.

**Comment by npitre** [comment]
> No doubt lots of bikeshedding will happen over the feature removal

Fear no more. The bikeshed has retained all its features with the last push.  ;-)

**Review by andyross — DISMISSED** [review]
Looks as clean as this hardware is as likely to get.

File: drivers/timer/apic_tsc.c [hunk 2]
```diff
@@ -39,63 +101,103 @@ static ALWAYS_INLINE uint64_t rdtsc(void)
...
-	uint32_t lo = (uint32_t) val;
+		set_trigger(next_cycle);
+	}
 
-	__asm__ volatile("wrmsr" :: "d"(hi), "a"(lo), "c"(msr));
+	k_spin_unlock(&lock, key);
+	sys_clock_announce(delta_ticks);
 }
 
 void sys_clock_set_timeout(int32_t ticks, bool idle)
```
Comment (kwd-doodling) [inline drivers/timer/apic_tsc.c:151]: one issue captured on ISH here, there's a chance that sys_clock_set_timeout() is called with ticks=0. Then ICR is programmed with zero, so the timer stops.

**Comment by kwd-doodling** [comment]
@npitre this patch does enhance apic_tsc.c, look good 👋

File: drivers/timer/apic_tsc.c (hunk 2 above)
Comment (npitre) [reply drivers/timer/apic_tsc.c:151]: Fixed.

**Review by kwd-doodling — APPROVED** [review]

**Review by edersondisouza — APPROVED** [review]

**Pull request merged**